// dbus-ts/lib/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultAuthMethods = exports.messageSignature = exports.endianness = exports.flags = exports.protocolVersion = exports.headerTypeId = exports.fieldSignature = exports.headerTypeName = exports.messageType = void 0;
exports.messageType = {
    invalid: 0,
    methodCall: 1,
    methodReturn: 2,
    error: 3,
    signal: 4
}, exports.headerTypeName = [
    null,
    'path',
    'interface',
    'member',
    'errorName',
    'replySerial',
    'destination',
    'sender',
    'signature'
], 
// TODO: merge to single hash? e.g path -> [1, 'o']
exports.fieldSignature = {
    path: 'o',
    interface: 's',
    member: 's',
    errorName: 's',
    replySerial: 'u',
    destination: 's',
    sender: 's',
    signature: 'g'
}, exports.headerTypeId = {
    path: 1,
    interface: 2,
    member: 3,
    errorName: 4,
    replySerial: 5,
    destination: 6,
    sender: 7,
    signature: 8
}, exports.protocolVersion = 1, exports.flags = {
    noReplyExpected: 1,
    noAutoStart: 2
}, exports.endianness = {
    le: 108,
    be: 66
}, exports.messageSignature = 'yyyyuua(yv)', exports.defaultAuthMethods = ['EXTERNAL', 'DBUS_COOKIE_SHA1', 'ANONYMOUS'];
